README.md
# clojureBNT
my notes for clojure for the brave and true


fwpd/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-02-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-02-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/fwpd/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/fwpd/compare/0.1.0...0.1.1


ch4.clj
(ns clojureBNT.ch4
  (:require [clojure.string :as string]))

;; Core Functions In Depth
; Programming to Abstractions
; - abstractions~ named collections of operations
; - ie: map doesn't care about collection, it cares about operation on collection

; Treating Lists, Vectors, Sets, and Maps as Seqs
; - Seqs are linearly ordered collections of elements
; - constrasts with unordered or graph ordered collections
; - operations that take advantage of this are said to "implement" seq abstraction
; - example operations include...

(defn titleize
  [topic]
  (str topic " for the brave and true"))

(map titleize ["Hamsters" "Ragnarok"])
(map titleize '("Empathy" "Decorating"))
(map titleize #{"Elobws" "soap carving"})
(map #(titleize (second %)) {:uncomfortable-thing "Winking"})

; first, rest and cons -> concepts in JS-like
#_(Linked Lists in generic terms
          var node3 = {
                       value "last"
                       next null}
          var node2 = {
                       value "middle"
                       next node3}
          var node1 = {
                       value "first"
                       next node2}

          var first = function (array) {
                                        return array[0]
                                        }
          var rest = function (array) {
                                       var sliced = array.slice (1, array.length)
                                       if (sliced.length==0) {
                                                              return null
                                                              }
                                       else {
                                             return sliced;
                                             }
                                       }
          var cons = function (newValue, array) {
                                                 return [newValue].concat (array)
                                                 }

          (first node1 => "first"
                 first (rest (node1)) => "middle"
                 first (rest (rest (node1))) => "last"
                 var node0 = cons ("new first", node1)
                 first (node0) => "new first"
                 first (rest (node0)) => "first")

          #(var map = function (list, transform) {
                                                  if (list == null) {
                                                                     return null}
                                                  else {
                                                        return cons (transform (first (list)), map (rest (list), transform))
                                                        }
                                                  }))

; abstraction through indirection
; - indirection: mechanisms to employ multiple related meanings
; - - polymorphism: mechanism to disbatch to different function bodies based on arg type
; - - before map/first/rest/cons: clojure calls seq on the collection
(seq '(1 2 3))
(seq [1 2 3])
(seq #{ 1 2 3})

(seq {:name "Bill Compton" :occupation "dead mopey guy"})
(into {} (seq {:a 1 :b 2 :c 3}))

;; Seq Function Examples
; map
(map inc [1 2 3])
(map str ["a" "b" "c"] ["A" "B" "C"])
(list (str "a" "A") (str "b" "B") (str "c" "C"))

(def human-consumption [8.1 7.3 6.6 5.0])
(def critter-consumption [0.0 0.2 0.3 1.1])
(defn unify-diet-data
  [human critter]
  {:human human
   :critter critter})
(map unify-diet-data human-consumption critter-consumption)

(def sum #(reduce + %))
(def avg #(/ (sum %) (count %)))
(defn stats
  [numbers]
  (map #(% numbers) [sum count avg]))
(stats [3 4 10])
(stats [80 1 44 13 6])

(def identities
  [{:alias "Batman" :real "Bruce Wayne"}
   {:alias "Spider-man" :real "Peter Parker"}
   {:alias "Santa" :real "Your mom"}
   {:alias "Easter Bunny" :real "Your dad"}])
(map :real identities)

; reduce
(reduce (fn [new-map [key val]]
          (assoc new-map key (inc val)))
        {}
        {:max 30 :min 10})
(assoc (assoc {} :max (inc 30))
       :min (inc 10))

 (reduce (fn [new-map [key val]]
           (if (> val 4)
             (assoc new-map key val)
             new-map))
         {}
         {:human 4.1
          :critter 3.9})

; take, drop, take-while, drop-while
(take 3 [1 2 3 4 5 6 7 8 9 10])
(drop 3 [1 2 3 4 5 6 7 8 9 10])

(def food-journal
  [{:month 1 :day 1 :human 5.3 :critter 2.3}
   {:month 1 :day 2 :human 5.1 :critter 2.0}
   {:month 2 :day 1 :human 4.9 :critter 2.1}
   {:month 2 :day 2 :human 5.0 :critter 2.5}
   {:month 3 :day 1 :human 4.2 :critter 3.3}
   {:month 3 :day 2 :human 4.0 :critter 3.8}
   {:month 4 :day 1 :human 3.7 :critter 3.9}
   {:month 4 :day 2 :human 3.7 :critter 3.6}])
(take-while #(< (:month %) 3) food-journal)
(drop-while #(< (:month %) 3) food-journal)
(take-while #(< (:month %) 4)
            (drop-while #(< (:month %) 2) food-journal))

; filter and some
(filter #(< (:human %) 5) food-journal)
(filter #(< (:month %) 3) food-journal)
(some #(> (:critter %) 5) food-journal)
(some #(> (:critter %) 3) food-journal)
(some #(and (> (:critter %) 3) %) food-journal)

; sort and sort-by
(sort [1 3 2])
(sort-by count ["aaa" "c" "bb"])

; concat
(concat [1 2] [2 4])

;; Lazy Seqs
; demonstrating lazy seq efficiency
(def vampire-database
  {0 {:makes-blood-puns? false :has-pulse? true :name "McFishwich"}
   1 {:makes-blood-puns? false :has-pulse? true :name "McMackson"}
   2 {:makes-blood-puns? true :has-pulse? false :name "Damon Salvatore"}
   3 {:makes-blood-puns? true :has-pulse? true :name "Mickey Mouse"}})

(defn vampire-related-details
  [ssn]
  (Thread/sleep 1000)
  (get vampire-database ssn))

(defn vampire?
  [record]
  (and (:makes-blood-puns? record)
       (not (:has-pulse? record))
       record))

(defn identify-vampire
  [ssns]
  (first (filter vampire?
                 (map vampire-related-details ssns))))

(time (vampire-related-details 0))
(time (def mapped-details (map vampire-related-details (range 0 1000000))))
(time (first mapped-details))
(time (first mapped-details))
(time (identify-vampire (range 0 1000000)))

; infinite sequences
(concat (take 8 (repeat "na")) ["Batman!"])

(take 3 (repeatedly (fn [] (rand-int 10))))

(defn even-numbers
  ([] (even-numbers 0))
  ([n] (cons n (lazy-seq (even-numbers (+ n 2))))))
(take 10 (even-numbers))
(cons 0 '(2 4 6))

;; Collection Abstraction
; - Similar to seq, also obeyed by most data structures (sets, maps, lists, vectors)
; - Collection is principle of apply operation to data structure as a whole
(empty? [])
(empty? ["no!"])

; into
(map identity {:sunlight-reaction "Glitter!"})
(into {} (map identity {:sunlight-reaction "Glitter!"}))

(map identity [:garlic :sesame-oil :fried-eggs])
(into [] (map identity [:garlic :sesame-oil :fried-eggs]))
(into #{} (map identity [:garlic :garlic]))
(into {:favorite-emotion "gloomy"} [[:sunlight-reaction "Glitter!"]])

; conj
(conj [0] [1])
(into [0] [1])
(conj [0] 1)
(conj [0] 1 2 3 4)
(conj {:time "midnight"} [:place "ye olde cemetarium"])
(defn my-conj
  [target & additions]
  (into target additions))
(my-conj [0] 1 2 3)

;; Function Functions
; apply
(max 0 1 2)
(max [0 1 2])
(apply max [1 2 3])

(defn my-into
  [target additions]
  (apply conj target additions))
(my-into [0]  [1 2 3])

; partial
(def add10 (partial + 10))
(add10 3)
(add10 5)

(def add-missing-elements
  (partial conj ["water" "earth" "air"]))
(add-missing-elements "fire")

(defn my-partial
  [partialized-fn & args]
  (fn [& more-args]
    (apply partialized-fn (into args more-args))))
(def add20 (my-partial + 20))
(add20 43)
(fn [& more-args]
  (apply + (into [20] more-args)))

(defn lousy-logger
  [log-level message]
  (condp = log-level
    :warn (clojure.string/lower-case message)
    :emergency (clojure.string/upper-case message)))

(def warn (partial lousy-logger :warn))
(warn "red light AHEAD")

; complement
(defn indentify-humans
  [ssns]
  (filter #(not (vampire? %))
          (map vampire-related-details ssns)))

(def not-vampire? (complement vampire?))
(defn identify-humans
  [ssns]
  (filter not-vampire? (map vampire-related-details ssns)))

(comment
  (defn my-complement
    [fun]
    (fn [& args]
      (not (apply fun args)))))

(def my-pos? (complement neg?))
(my-pos? 1)
(my-pos? -1)

;; Vampire Data Analysis Program for FWPD
; - See fwpd app

;; Exercises
; 1. Turn the result of your glitter filter into a list of names.
; - see fwpd
; 2. Write a function, append, which will append a new suspect to your list of suspects.
(defn append
  "adds a suspect to a list of suspects"
  [new-suspect suspect-list]
  (into suspect-list new-suspect))
; => new suspects should look like: {# {:makes-blood-puns? boolean, :has-pulse? boolean, :name "string"}}

; 3. Write a function, validate, which will check name and glitter-index when appending.
;   - should take a map of keywards and record to be validated
(defn validate
  [fn-keywords record]
  (= (keys record) (keys fn-keywords)))

; 4. Write a function that takes a list of maps and converts to a csv string using string/join
(defn convert-csv
  [map-list]
  (let [stringed-list (clojure.string/join "," (map identity map-list))]
       (clojure.string/replace (clojure.string/replace (stringed-list) "{" "") "}" "")))



fwpd/src/fwpd/core.clj
(ns fwpd.core)
(def filename "suspects.csv")

(slurp filename)

(def vamp-keys [:name :glitter-index])

(defn str->int
  [str]
  (Integer. str))

(def conversions {:name identity
                  :glitter-index str->int})

(defn convert
  [vamp-key value]
  ((get conversions vamp-key) value))

(defn parse
  "convert csv to rows of cols"
  [string]
  (map #(clojure.string/split % #",")
       (clojure.string/split string #"\n")))

(parse (slurp filename))

(defn mapify
  "return a seq of maps"
  [rows]
  (map (fn [unmapped-row]
         (reduce (fn [row-map [vamp-key value]]
                   (assoc row-map vamp-key (convert vamp-key value)))
                 {}
                 (map vector vamp-keys unmapped-row)))
       rows))

(first (mapify (parse (slurp filename))))

(defn glitter-filter
  [minimum-glitter records]
  (filter #(>= (:glitter-index %) minimum-glitter) records))

(glitter-filter 3 (mapify (parse (slurp filename))))

; exercise 1
(defn name-glitter-filter
  [minimum-glitter records]
  (map :name (glitter-filter minimum-glitter records)))

(name-glitter-filter 3 (mapify (parse (slurp filename))))




ch3.clj
(ns clojureBNT.ch3
  (:require [clojure.string :as string]))

;; syntax
; forms
"a string"
["a" "vector" "of" "strings"]

; operators
(comment
  (operator operand1 operand2 ... operandn))
(+ 1 2 3)
(str "It was panda " "in library " "with dust buster")

; control flow
; if
(comment
  (if boolean-form
    then-form
    optional-else-form))
(if true
  "By Zeus's hammer!"
  "By aquaman's trident!")
(if false
  "By Zeus's hammer!"
  "By aquaman's trident!")
(if false
  "By odin's elbow!")

; do
(if true
  (do (println "do success!")
      "By Zeus's hammer!")
  (do (println "do Failure!")
      "By Aquaman's Trident!"))

; when
(when true
  (println "when success")
  "abra cadabra")

; nil, true, false, truthiness, equality, boolean exps
(nil? 1)
(nil? nil)

(if "bears eat beets"
  "Battlestar Galactica")
(if nil
  "This won't be result"
  "nil is falsey")

(= 1 1)
(= nil nil)
(= 1 2)

(or false nil :large_I_mean_venti :why_cant_I_say_large)
(or (= 0 1) (= "yes" "no"))
(or nil)

(and :free_wifi :hot_coffee)
(and :feelin_super_cool nil false)

;; Naming Values with Def
; def
(def failed-protagonist-names
  ["Larry Potter" "Doreen the Explorer" "The Incredible Bulk"])
failed-protagonist-names

(defn error-message
  [severity]
  (str "Oh god, it's a disaster, we're "
       (if (= severity :mild)
         "mildly inconvenienced"
         "doooooomed")))

(error-message :mild)
(error-message :anything)

;; Data structures
; numbers
93
1.2
1/5

; strings
"Lord voldemort"
"\"He who must not be named\""
"\"Great cow of Moscow!\" - Hermes Conrad"

(def name "Chewbacca")
(str "\"Ugglglg\" - " name)

;maps
{}
{:first-name "charlie"
 :last-name "McFishwich"}

{"string-key" +}

{:name {:first "John"
        :middle "Jacob"
        :last "Jingleheimerschmit"}}

(hash-map :a 1 :b 2)

(get {:a 0 :b 1} :b)
(get {:a 0 :b {:c "ho hum"}} :b)
(get {:a 0 :b 1} :c)
(get {:a 0 :b 1} :c "unicorns?")

(get-in {:a 0 :b {:c "ho-hum"}} [:b :c])

({:name "The human coffeepot"} :name)

; keywords
:a
:rumbplestilsken
:34
:_?

(:a {:a 1 :b 2 :c 3})

(get {:a 1 :b 2 :c 3} :a)

(:d {:a 1 :b 2 :c 3} "no gnome knows homes like noah knows")

; vectors
[3 2 1]

(get [3 2 1] 0)
(get ["a" {:name "pubsley winterbottom"} "c"] 1)
(vector "creepy" "full" "moon")

(conj [1 2 3] 4)

; lists
'(1 2 3 4)

(nth '(:a :b :c) 0)
(nth '(:a :b :c) 2)

(list 1 "two" {3 4})
(conj '(1 2 3) 4)

; sets
#{"kurt vonnegut" 20 :icicle}

(hash-set 1 1 2 2)

(conj #{:a :b} :b)

(set [3 3 3 4 4])

(contains? #{:a :b} :a)
(contains? #{:a :b} 3)
(contains? #{nil} nil)

(:a #{:a :b})

(get #{:a :b} :a)
(get #{:a nil} nil)
(get #{:a :b} "kur vonnegut")


;; simplicity
; better to have 100 functions on one data struct
; than 10 functions on 10 structures

;; functions
; calling functions
(+ 1 2 3 4)
(* 1 2 3 4)
(first [1 2 3 4])

((or + -) 1 2 3)
((and (= 1 1) +) 1 2 3)
((first [+ 0]) 1 2 3)

(comment
  "incorrect func calls"
  (1 2 3 4)
  ("test" 1 2 3))

(inc 1.1)
(map inc [0 1 2 3])

(+ (inc 199) (/ 100 (- 7 2)))

;; function, macros special forms
(comment
  (if good-mood
    (tweet walking-on-sunshine-lyrics)
    (tweet mopey-country-song-lyrics))
  )

; defining functions
(defn too-enthusiastic
  "return a cheer that is too much"
  [name]
  (str "OMG " name " you are the best"))

(too-enthusiastic "Zelda")

; docstring:
(comment
  docstring returns when calling:
  (doc fn-name))

; parameters and arity
(defn no-params
  []
  "I take no parameters")

(defn one-param
  [x]
  (str "I take one param: " x))

(defn two-params
  [x y]
  (str "smushing too params" x y))

(comment (defn multi-arity
           ([first-arg second-arg third-arg]
            (do-things first-arg second-arg third-arg))
           ([first-arg second-arg]
            (do-things first-arg second-arg))
           ([first-arg]
            (do-things first-arg))))

(defn x-chop
  "describr the kind of chop your using"
  ([name chop-type]
   (str "I " chop-type " chop " name))
  ([name]
   (x-chop name "karate")))

(x-chop "kanye west" "slap")
(x-chop "kanye east")

(defn weird-arity
  ([]
   "docstring with zero-arity")
  ([number]
   (inc number)))

(defn codger-communication
  [whippersnapper]
  (str "Get off my lawn, " whippersnapper "!!!"))

(defn codger
  [& whippersnappers]
  (map codger-communication whippersnappers))

(codger "Biller" "Anne-Marie" "The Incredible Hulk")

(defn favorite-things
  [name & things]
  (str "Hi, " name ", here are my favorite things:"
       (clojure.string/join ", " things)))

(favorite-things "Doreen" "gum" "shoes" "kara-te")

; destructuring
(defn my-first
  [[first-thing]]
  first-thing)

(my-first ["oven" "bike" "war-axe"])

(defn chooser
  [[first-choice second-choice & unimportant-choices]]
  (println (str "Your first choice is: " first-choice))
  (println (str "Your second choice is: " second-choice))
  (println (str "We're ignoring the rest of your choices. "
                "Here they are in case you need them: "
                (clojure.string/join "," unimportant-choices))))

(chooser ["Marmalade", "Handsome Jack", "Pigpen", "Aquaman"])

(defn announce-treasure-location
  [{lat :lat lng :lng}]
  (println (str "Treasure lat: " lat))
  (println (str "treasure lng: " lng)))

(announce-treasure-location {:lat 28.22 :lng 81.33})

; function body
(defn illustrative-function
  []
  (+ 1 304)
  30
  "joe")

(illustrative-function)

(defn number-comment
  [x]
  (if (> x 6)
    "Oh my gosh! what a big number!"
    "That number's OK, I guess"))

(number-comment 5)
(number-comment 7)

; all functions are created equal

; Anonymous Functions
(comment (fn [param-list]
           function body))

(map (fn [name] (str "hi, " name))
     ["darth vader" "mr. magoo"])

((fn [x] (* x 3)) 8)

(def my-special-multiplier (fn [x] (* x 3)))
(my-special-multiplier 12)

#(* % 3)

(#(* % 3) 8)

(map #(str "Hi, " %)
     ["Darth vader" "mr. magoo"])

(* 8 3)
#(* % 3)

(#(str %1 " and " %2) "cornbread" "butter beans")
(#(identity %&) 1 "blarg" :yip)

; returning functions
(defn inc-maker
  [inc-by]
  #(+ % inc-by))
(def inc3 (inc-maker 3))
(inc3 7)

; Shire's Next Top Model
 (def asym-hobbit-body-parts [{:name "head" :size 3}
                             {:name "left-eye" :size 1}
                             {:name "left-ear" :size 1}
                             {:name "mouth" :size 1}
                             {:name "nose" :size 1}
                             {:name "neck" :size 2}
                             {:name "left-shoulder" :size 3}
                             {:name "left-arm" :size 3}
                             {:name "chest" :size 10}
                             {:name "back" :size 10}
                             {:name "left-forearm" :size 3}
                             {:name "abdomen" :size 6}
                             {:name "left-kidney" :size 1}
                             {:name "left-hand" :size 2}
                             {:name "left-knee" :size 2}
                             {:name "left-thigh" :size 4}
                             {:name "left-leg" :size 3}
                             {:name "left-achilles" :size 1}
                             {:name "left-foot" :size 2}])

(defn matching-part
  [part]
  {:name (clojure.string/replace (:name part) #"^left-" "right-")
   :size (:size part)})

(defn symmetrize-body-parts
  "expects seq of maps that have a :name and :size"
  [asym-body-parts]
  (loop [remaining-asym-parts asym-body-parts
         final-body-parts []]
    (if (empty? remaining-asym-parts)
      final-body-parts
      (let [[part & remaining] remaining-asym-parts]
        (recur remaining
               (into final-body-parts
                     (set [part (matching-part part)])))))))

(symmetrize-body-parts asym-hobbit-body-parts)

; let
(let [x 3]
  x)

(def dalmatian-list
  ["pongo" "perdita" "puppy 1" "puppy 2"])
(let [dalmatians (take 2 dalmatian-list)]
  dalmatians)

(def x 0)
(let [x 1]
  x)
x
(def x 0)
(let [x (inc x)]
  x)

(let [[pongo & dalmatians] dalmatian-list]
  [pongo dalmatians])

; loop
(loop [iteration 0]
  (println (str "interation " iteration))
  (if (> iteration 3)
    (println "Goodbye")
    (recur (inc iteration))))

(defn recursive-printer
  ([]
   (recursive-printer 0))
  ([iteration]
   (println iteration)
   (if (> iteration 3)
     (println "goodbye")
     (recursive-printer (inc iteration)))))
(recursive-printer)

; regular expressions
#"regular-expression"
(re-find #"^left-" "left-eye")
(re-find #"^left-" "cleft-chin")
(re-find #"^left-" "other right")

; better symmetrizer with reduce:
(reduce + [1 2 3])
(+ (+ 1 2) 3)
(reduce + 10 [1 2 3])

(defn my-reduce
  ([f initial coll]
   (loop [result initial
          remaining coll]
     (if (empty? remaining)
       result
       (recur (f result (first remaining)) (rest remaining)))))
  ([f [head & tail]]
   (my-reduce f head tail)))

(defn better-symmetrize-body-parts
  "expects seq of maps that have a :name and :size"
  [asym-body-parts]
  (reduce (fn [final-body-parts part]
            (into final-body-parts (set [part (matching-part part)])))
          []
          asym-body-parts))

; hobbit violence:
(defn hit
  [asym-body-parts]
  (let [sym-parts (better-symmetrize-body-parts asym-body-parts)
        body-part-size-sum (reduce + (map :size sym-parts))
        target (rand body-part-size-sum)]
    (loop [[part & remaining] sym-parts
           accumulated-size (:size part)]
      (if (> accumulated-size target)
        part
        (recur remaining (+ accumulated-size (:size (first remaining))))))))

(hit asym-hobbit-body-parts)
(hit asym-hobbit-body-parts)

;; Exercises:
; 1. Use str, vector, list, hash-map, hash-set
(str "hey " "there")
(vector 1 2 3)
(list 1 2 3)
(hash-map :a 1 :b 2 :c 3)
(hash-set 1 1 2 3 3)
; 2. Write a function that takes a number and adds 100
(defn hundred-adder
  [number]
  (+ number 100))
(hundred-adder 5)
; 3. Write a function dec-maker that works like inc-maker but with dec
(defn dec-maker
  [dec-val]
  #(- % dec-val))
((dec-maker 5) 86)
; 4. Write a funciton mapset that works like map except returns a set
(comment
  (mapset inc [1 1 2 2])
  => #{2 3})
(defn mapset
  [coll]
  (set (map inc coll)))
(mapset [1 1 2 2])
; 5. Create a function similar to sym-body-parts, but for 5-parted radial
(defn matching-radial
  [part]
  [{:name (clojure.string/replace (:name part) #"^left-" "middle-")
    :size (:size part)}
   {:name (clojure.string/replace (:name part) #"^left-" "top-")
    :size (:size part)}
   {:name (clojure.string/replace (:name part) #"^left-" "right-")
   :size (:size part)}
   {:name (clojure.string/replace (:name part) #"^left-" "bottom-")
   :size (:size part)}])

(matching-radial {:name "left-hand" :size 3})

(defn radialize-body-parts
  "expects seq of maps that have a :name and :size"
  [asym-body-parts]
  (loop [remaining-asym-parts asym-body-parts
         final-body-parts []]
    (if (empty? remaining-asym-parts)
      (distinct (flatten final-body-parts))
      (let [[part & remaining] remaining-asym-parts]
        (recur remaining
               (into final-body-parts
                     (set [part (matching-radial part)])))))))

(clojure.pprint/pprint (radialize-body-parts asym-hobbit-body-parts))

; 6. Create a function that generalizes sym-body-parts
(defn generlize-body-parts
  [asym-body-parts gen-count]
  )
